app/controllers/static_pages_controller.rb

```diff
@@ -1,3 +1,6 @@
 class StaticPagesController < ApplicationController
-  def home; end
+  def home
+    @products = Product.includes(:images).all.page(params[:page]).per(Settings
```

Comment:
includes khác gì vs eger load vs preload e ?

Reply:
có cần .all k

Reply:
includes, egerload ,preload là 3 method khi lấy dữ liệu tránh N+1 query. 
3 khác nhau chỉ ở các query dữ liệu từ db. thường thì dùng join hoặc subquery 

-đầu tiên includes: 2 câu query để lấy product và image
  Product Load (0.6ms)  SELECT `products`.* FROM `products`
  Image Load (0.8ms)  SELECT `images`.* FROM `images` WHERE `images`.`product_id` IN (1, 2, 3, 4, 5, 6, 7, 8,..)

- eager_load: 1 câu query, sử dụng left outer join để tạo 1 bảng tạm chứa dữ liệu cần lấy

Product.eager_load(:images).all
   SQL (1.6ms) SELECT `products`.`id` AS t0_r0, `products`.`name` AS t0_r1, `products`.`description` AS t0_r2, `products`.`price` AS t0_r3, `products`.`inventory_number` AS t0_r4, `products`.`category_id` AS t0_r5, `products`.`supplier_id` AS t0_r6, `products`.`created_at` AS t0_r7, `products`.`updated_at` AS t0_r8, `images`.`id` AS t1_r0, `images`.`product_id` AS t1_r1, `images`.`image_data` AS t1_r2, `images`.`created_at` AS t1_r3, `images`.`updated_at` AS t1_r4 FROM `products` LEFT OUTER JOIN `images` ON `images`.`product_id` = `products`.`id` 

-preload: 2 câu query giống includes nhưng thời gian chạy lâu hơn:
  Product Load (0.8ms)  SELECT `products`.* FROM `products`
  Image Load (1.0ms)  SELECT `images`.* FROM `images` WHERE `images`.`product_id` IN (1, 2, 3, 4, 5, 6, 7, 8, 9, 10..)

=> kết luận dụng includes sẽ có tốc độ truy vấn tốn hơn 2 thằng còn lại
-nếu muốn dữ liệu lấy ở bảng liên kết với nó nhưng dùng includes thì dùng includes().references() nó sẽ sinh ra query join những bảng lại.